wagtail/core/models.py

```diff
@@ -3185,6 +3538,201 @@ def get_comment(self):
+        next_task_data = None
+        if next_task:
+            next_task_data = {
+                'id': next_task.id,
+                'title': next_task.name
+            }
+        PageLogEntry.objects.log_action(
+            instance=page,
+            action='wagtail.workflow.{}'.format(action),
+            user=user,
+            data={
+                'workflow': {
+                    'id': self.workflow_state.workflow.id,
+                    'title': self.workflow_state.workflow.name,
+                    'status': self.status,
+                    'task_state_id': self.id,
+                    'task': {
+                        'id': self.task.id,
+                        'title': self.task.name,
+                    },
+                    'next': next_task_data,
+                },
+                'comment': self.get_comment()
+            },
+            revision=self.page_revision
+        )
+
     class Meta:
         verbose_name = _('Task state')
         verbose_name_plural = _('Task states')
+
+
+class BaseLogEntryManager(models.Manager):
+    def log_action(self, instance, action, **kwargs):
+        """
+        :param instance: The model instance we are logging an action for
+        :param action: The action. Should be namespaced to app (e.g. wagtail.create, wagtail.workflow.start)
+        :param kwargs: Addition fields to for the model deriving from BaseLogEntry
+            - user: The user performing the action
+            - title: the instance title
+            - data: any additional metadata
+            - content_changed, deleted - Boolean flags
+        :return: The new log entry
+        """
+        data = kwargs.pop('data', '')
+        title = kwargs.pop('title', None)
+        if not title:
+            if hasattr(instance, 'get_admin_display_title'):
+                title = instance.get_admin_display_title()
+            else:
+                title = str(instance)
+
+        timestamp = kwargs.pop('timestamp', timezone.now())
+        return self.model.objects.create(
+            content_type=ContentType.objects.get_for_model(instance, for_concrete_model=False),
+            label=title,
+            action=action,
+            timestamp=timestamp,
+            data_json=json.dumps(data),
+            **kwargs,
+        )
+
+    def get_for_model(self, model):
+        # Return empty queryset if the given object is not valid.
+        if not issubclass(model, models.Model):
+            return self.none()
+
+        ct = ContentType.objects.get_for_model(model)
+
+        return self.filter(content_type=ct)
+
+    def get_for_user(self, user_id):
+        return self.filter(user=user_id)
+
+
+class PageLogEntryManager(BaseLogEntryManager):
+
+    def log_action(self, instance, action, **kwargs):
+        kwargs.update(page=instance)
+        return super().log_action(instance, action, **kwargs)
+
+
+class BaseLogEntry(models.Model):
+    content_type = models.ForeignKey(
```

Comment:
Is this field now redundant since we have type-specific `LogEntry` models?

Reply:
This is useful when the page is deleted and `object_verbose_name` uses it